native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/NativeObjCNameChecker.kt

```diff
@@ -0,0 +1,115 @@
+                context.trace.report(ErrorsNative.INAPPLICABLE_OBJC_NAME.on(reportLocation))
+            }
+        }
+        objCNames.forEach { checkObjCName(it, declaration, descriptor, context) }
+    }
+
+    private fun checkObjCName(
+        objCName: ObjCName,
+        declaration: KtDeclaration,
+        descriptor: DeclarationDescriptor,
+        context: DeclarationCheckerContext
+    ) {
+        if (objCName.name == null && objCName.swiftName == null) {
+            val reportLocation = DescriptorToSourceUtils.getSourceFromAnnotation(objCName.annotation) ?: declaration
+            context.trace.report(ErrorsNative.INVALID_OBJC_NAME.on(reportLocation))
+        }
+        val invalidNameChars = objCName.name?.toSet()?.intersect(NativeIdentifierChecker.invalidChars) ?: emptySet()
+        val invalidSwiftNameChars = objCName.swiftName?.toSet()?.intersect(NativeIdentifierChecker.invalidChars) ?: emptySet()
+        val invalidChars = invalidNameChars + invalidSwiftNameChars
+        if (invalidChars.isNotEmpty()) {
+            val reportLocation = DescriptorToSourceUtils.getSourceFromAnnotation(objCName.annotation) ?: declaration
+            context.trace.report(ErrorsNative.INVALID_CHARACTERS_OBJC_NAME.on(reportLocation, invalidChars.joinToString("")))
+        }
+        if (objCName.exact && descriptor !is ClassDescriptor) {
+            val reportLocation = DescriptorToSourceUtils.getSourceFromAnnotation(objCName.annotation) ?: declaration
+            context.trace.report(ErrorsNative.INAPPLICABLE_EXACT_OBJC_Name.on(reportLocation))
+        }
+    }
+
+    private fun checkOverrides(declaration: KtDeclaration, descriptor: DeclarationDescriptor, context: DeclarationCheckerContext) {
+        if (descriptor !is CallableMemberDescriptor || descriptor.overriddenDescriptors.isEmpty()) return
+        val objCNames = descriptor.overriddenDescriptors.map { it.getFirstBaseDescriptor().getObjCNames() }
+        if (!objCNames.allNamesEquals()) {
+            context.trace.report(ErrorsNative.INCOMPATIBLE_OBJC_NAME_OVERRIDE.on(declaration))
+        }
+    }
+
+    private fun CallableMemberDescriptor.getFirstBaseDescriptor(): CallableMemberDescriptor =
+        if (overriddenDescriptors.isEmpty()) this else overriddenDescriptors.first().getFirstBaseDescriptor()
+
+    private class ObjCName(
+        val annotation: AnnotationDescriptor
+    ) {
+        val name: String? = annotation.argumentValue("name")?.value?.cast<String>()?.takeIf { it.isNotBlank() }
+        val swiftName: String? = annotation.argumentValue("swiftName")?.value?.cast<String>()?.takeIf { it.isNotBlank() }
+        val exact: Boolean = annotation.argumentValue("exact")?.value?.cast() ?: false
+
+        override fun equals(other: Any?): Boolean =
+            other is ObjCName && name == other.name && swiftName == other.swiftName && exact == other.exact
+
+        override fun hashCode(): Int {
+            var result = name?.hashCode() ?: 0
```

Comment:
This extension function might be useful here and below: https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/hash-code.html